README.md
# King-Mammoth-Installer
The Official King Mammoth Installer


releases
King Mammoth GAMMA Forge 0.9.8.zip,king-mammoth.org/releases/King-Mammoth-GAMMA-Forge-0.9.8.zip,05/17/19


src/test/java/king_mammoth_org/King_Mammoth_Launcher/AppTest.java
package king_mammoth_org.King_Mammoth_Launcher;

import junit.framework.Test;
import junit.framework.TestCase;
import junit.framework.TestSuite;

/**
 * Unit test for simple App.
 */
public class AppTest 
    extends TestCase
{
    /**
     * Create the test case
     *
     * @param testName name of the test case
     */
    public AppTest( String testName )
    {
        super( testName );
    }

    /**
     * @return the suite of tests being tested
     */
    public static Test suite()
    {
        return new TestSuite( AppTest.class );
    }

    /**
     * Rigourous Test :-)
     */
    public void testApp()
    {
        assertTrue( true );
    }
}


src/main/java/king_mammoth_org/kingmammothlauncher/swing/Release.java
package king_mammoth_org.kingmammothlauncher.swing;

import java.io.File;
import java.io.FileOutputStream;
import java.io.IOException;
import java.net.URL;
import java.nio.channels.Channels;
import java.nio.channels.ReadableByteChannel;
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.Paths;
import java.util.List;

public class Release {

	String name;
	String url;
	String date;

	public Release(String name, String url, String date) {

		this.name = name;
		this.url = url;
		this.date = date;

	}

	public static Release[] getReleases(File file) throws IOException {

		List<String> lines = Files.readAllLines(Paths.get(file.getAbsolutePath()), StandardCharsets.UTF_8);
		Release[] releases = new Release[lines.size()];

		for (int i = 0; i < releases.length; i++) {

			String[] data = lines.get(i).split(",");

			Release release = new Release(data[0], data[1], data[2]);
			releases[i] = release;

		}

		return releases;

	}

	public static File getReleaseFile() throws IOException {

		String cwd = System.getProperty("user.dir");

		URL website = new URL("http://king-mammoth.org//releases//releases.txt");
		ReadableByteChannel rbc = Channels.newChannel(website.openStream());
		FileOutputStream fos = new FileOutputStream(cwd + "/releases.txt");
		fos.getChannel().transferFrom(rbc, 0, Long.MAX_VALUE);
		fos.close();

		return new File(cwd + "/releases.txt");

	}

}


src/main/java/king_mammoth_org/kingmammothlauncher/swing/HTTPDownloadUtil.java
package king_mammoth_org.kingmammothlauncher.swing;

import java.io.IOException;
import java.io.InputStream;
import java.net.HttpURLConnection;
import java.net.URL;

public class HTTPDownloadUtil {

	private HttpURLConnection httpConn;

	private InputStream inputStream;

	private String fileName;
	private int contentLength;

	public void downloadFile(String fileURL) throws IOException {
		URL url = new URL(fileURL);
		httpConn = (HttpURLConnection) url.openConnection();
		int responseCode = httpConn.getResponseCode();

		// always check HTTP response code first
		if (responseCode == HttpURLConnection.HTTP_OK) {
			String disposition = httpConn.getHeaderField("Content-Disposition");
			String contentType = httpConn.getContentType();
			contentLength = httpConn.getContentLength();

			if (disposition != null) {
				// extracts file name from header field
				int index = disposition.indexOf("filename=");
				if (index > 0) {
					fileName = disposition.substring(index + 10, disposition.length() - 1);
				}
			} else {
				// extracts file name from URL
				fileName = fileURL.substring(fileURL.lastIndexOf("/") + 1, fileURL.length());
			}

			// output for debugging purpose only
			System.out.println("Content-Type = " + contentType);
			System.out.println("Content-Disposition = " + disposition);
			System.out.println("Content-Length = " + contentLength);
			System.out.println("fileName = " + fileName);

			// opens input stream from the HTTP connection
			inputStream = httpConn.getInputStream();

		} else {
			throw new IOException("No file to download. Server replied HTTP code: " + responseCode);
		}
	}

	public void disconnect() throws IOException {
		inputStream.close();
		httpConn.disconnect();
	}

	public String getFileName() {
		return this.fileName;
	}

	public int getContentLength() {
		return this.contentLength;
	}

	public InputStream getInputStream() {
		return this.inputStream;
	}
}


src/main/java/king_mammoth_org/kingmammothlauncher/swing/EnchancedProgressBar.java
package king_mammoth_org.kingmammothlauncher.swing;

import javax.swing.JFrame;
import javax.swing.JProgressBar;
import javax.swing.JLabel;
import java.awt.Font;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.net.HttpURLConnection;
import java.net.URL;

import javax.swing.SwingConstants;
import javax.swing.SwingUtilities;
import javax.swing.SwingWorker;

import net.lingala.zip4j.ZipFile;
import net.lingala.zip4j.exception.ZipException;
import java.awt.Toolkit;

public class EnchancedProgressBar {

	public static JFrame frmDownloadModpack;
	public static JProgressBar progressBar;
	public static JLabel lblNewLabel_2;

	public static String url;
	public static String dir;
	
	public static long fileSize;
	public static long bytesDownloaded;
	public static int bytesRead = -1;

	/**
	 * Create the application.
	 * 
	 * @throws IOException
	 */

	public void start() {

		SwingUtilities.invokeLater(new Runnable() {
			@Override
			public void run() {

				frmDownloadModpack.setVisible(true);

			}
		});

	}

	public EnchancedProgressBar() throws IOException {
		initialize();
	}

	/**
	 * Initialize the contents of the frame.
	 * 
	 * @throws IOException
	 */

	static void initialize() throws IOException {
		frmDownloadModpack = new JFrame();
		frmDownloadModpack.setIconImage(Toolkit.getDefaultToolkit().getImage(EnchancedProgressBar.class.getResource("/king_mammoth_org/kingmammothlauncher/swing/download.png")));
		frmDownloadModpack.setTitle("Download Modpack");
		frmDownloadModpack.setBounds(100, 100, 450, 300);
		frmDownloadModpack.setDefaultCloseOperation(JFrame.EXIT_ON_CLOSE);
		frmDownloadModpack.getContentPane().setLayout(null);
		frmDownloadModpack.setResizable(false);

		progressBar = new JProgressBar();
		progressBar.setMinimum(0);
		progressBar.setBounds(43, 175, 350, 50);
		frmDownloadModpack.getContentPane().add(progressBar);

		JLabel lblNewLabel = new JLabel("Downloading modpack, please wait.");
		lblNewLabel.setHorizontalAlignment(SwingConstants.CENTER);
		lblNewLabel.setFont(new Font("Tahoma", Font.BOLD, 20));
		lblNewLabel.setBounds(0, 24, 434, 32);
		frmDownloadModpack.getContentPane().add(lblNewLabel);
		
		JLabel lblNewLabel_1 = new JLabel("<html> Please wait as we download the modpack and extract all files in the directory.");
		lblNewLabel_1.setHorizontalAlignment(SwingConstants.CENTER);
		lblNewLabel_1.setFont(new Font("Tahoma", Font.PLAIN, 15));
		lblNewLabel_1.setBounds(34, 64, 390, 50);
		frmDownloadModpack.getContentPane().add(lblNewLabel_1);
		
		lblNewLabel_2 = new JLabel("");
		lblNewLabel_2.setBounds(43, 150, 350, 14);
		frmDownloadModpack.getContentPane().add(lblNewLabel_2);

	}

	public void backgroundTask() {

		SwingWorker<Integer, String> sw1 = new SwingWorker<Integer, String>() {

			@Override
			protected Integer doInBackground() throws Exception {
				
				URL website = new URL(url);
				
				lblNewLabel_2.setText("Downloading Required Modpack Files");
				
				HttpURLConnection conn = (HttpURLConnection) website.openConnection();
				conn.setRequestMethod("HEAD");
				long fileSize = conn.getContentLengthLong();
				
				EnchancedProgressBar.fileSize = fileSize;
				progressBar.setMaximum((int)fileSize);
				
				HTTPDownloadUtil util = new HTTPDownloadUtil();
	            util.downloadFile(url);
				
				FileOutputStream fos = new FileOutputStream(dir + "/modpack.zip");
				InputStream inputStream = util.getInputStream();
				
				byte[] buffer = new byte[8096];

	            while ((bytesRead = inputStream.read(buffer)) != -1) {
	                fos.write(buffer, 0, bytesRead);
	                bytesDownloaded += bytesRead;
	                progressBar.setValue((int)bytesDownloaded);
	                
	            }
	             
				fos.close();
				
				return 1;
				
			}

			@Override
			protected void done() {
				
				lblNewLabel_2.setText("Extracting Zip File");
				
				ZipFile zipFile = new ZipFile(dir + "/modpack.zip");
				try {
					zipFile.extractAll(dir);
				} catch (ZipException e) {
					// TODO Auto-generated catch block
					e.printStackTrace();
				}
				
				frmDownloadModpack.setVisible(false);
				new Installed();
				
			}
		};

		sw1.execute();

	}
}


src/main/java/king_mammoth_org/kingmammothlauncher/swing/DownloadModpack.java
package king_mammoth_org.kingmammothlauncher.swing;

import java.awt.EventQueue;

import javax.swing.JFrame;
import javax.swing.JLabel;
import java.awt.Desktop;

import javax.swing.SwingConstants;
import javax.swing.UIManager;
import javax.swing.UnsupportedLookAndFeelException;

import java.awt.Font;
import javax.swing.JSeparator;
import javax.swing.JTextField;
import javax.swing.JList;
import javax.swing.JScrollPane;
import javax.swing.ListSelectionModel;
import javax.swing.JButton;
import javax.swing.JFileChooser;
import javax.swing.DefaultListModel;
import javax.swing.ImageIcon;
import java.awt.event.MouseAdapter;
import java.awt.event.MouseEvent;
import java.io.File;
import java.io.IOException;
import java.net.URI;
import java.net.URISyntaxException;
import java.awt.Color;
import java.awt.Toolkit;

public class DownloadModpack {

	int currentSelection = -1;

	private JFrame frmDownloadKingMammoth;
	private JTextField textField;

	File downloadDirectory;
	File fileToDownload;

	Release[] releases;

	/**
	 * Launch the application.
	 */
	public static void main(String[] args) {
		EventQueue.invokeLater(new Runnable() {
			public void run() {
				try {
					DownloadModpack window = new DownloadModpack();
					window.frmDownloadKingMammoth.setVisible(true);
				} catch (Exception e) {
					e.printStackTrace();
				}
			}
		});
	}

	/**
	 * Create the application.
	 */
	public DownloadModpack() {
		initialize();
	}

	/**
	 * Initialize the contents of the frame.
	 */
	private void initialize() {
		frmDownloadKingMammoth = new JFrame();
		frmDownloadKingMammoth.setIconImage(Toolkit.getDefaultToolkit().getImage(DownloadModpack.class.getResource("/king_mammoth_org/kingmammothlauncher/swing/launcher.png")));
		frmDownloadKingMammoth.setTitle("Download King Mammoth");
		frmDownloadKingMammoth.setBounds(100, 100, 450, 300);
		frmDownloadKingMammoth.setDefaultCloseOperation(JFrame.EXIT_ON_CLOSE);
		frmDownloadKingMammoth.getContentPane().setLayout(null);
		frmDownloadKingMammoth.setResizable(false);

		JLabel lblNewLabel = new JLabel("Download King Mammoth");
		lblNewLabel.setBounds(0, 5, 434, 41);
		lblNewLabel.setFont(new Font("Leelawadee UI", Font.BOLD, 30));
		lblNewLabel.setHorizontalAlignment(SwingConstants.CENTER);
		frmDownloadKingMammoth.getContentPane().add(lblNewLabel);

		JLabel lblNewLabel_1 = new JLabel(
				"<html> Please select your version to download, a download directory, and press Ok.");
		lblNewLabel_1.setFont(new Font("Tahoma", Font.PLAIN, 15));
		lblNewLabel_1.setBounds(10, 50, 414, 53);
		frmDownloadKingMammoth.getContentPane().add(lblNewLabel_1);

		JSeparator separator = new JSeparator();
		separator.setBounds(10, 50, 414, 2);
		frmDownloadKingMammoth.getContentPane().add(separator);

		JSeparator separator_1 = new JSeparator();
		separator_1.setBounds(10, 102, 414, 2);
		frmDownloadKingMammoth.getContentPane().add(separator_1);

		JLabel lblNewLabel_2 = new JLabel("Directory: ");
		lblNewLabel_2.setFont(new Font("Tahoma", Font.PLAIN, 10));
		lblNewLabel_2.setBounds(10, 114, 56, 14);
		frmDownloadKingMammoth.getContentPane().add(lblNewLabel_2);

		textField = new JTextField();
		textField.setBounds(70, 111, 255, 20);
		frmDownloadKingMammoth.getContentPane().add(textField);
		textField.setColumns(10);

		JSeparator separator_2 = new JSeparator();
		separator_2.setBounds(10, 139, 414, 2);
		frmDownloadKingMammoth.getContentPane().add(separator_2);

		DefaultListModel<String> model = new DefaultListModel<String>();

		File releaseFile = null;

		try {
			releaseFile = Release.getReleaseFile();
		} catch (IOException e4) {
			// TODO Auto-generated catch block
			e4.printStackTrace();
		}

		try {
			releases = Release.getReleases(releaseFile);

			for (int i = 0; i < releases.length; i++) {

				System.out.println(releases[i].name);
				System.out.println(releases[i].date);

				model.addElement(releases[i].name + "     | " + releases[i].date);

			}

		} catch (IOException e3) {
			// TODO Auto-generated catch block
			e3.printStackTrace();
		}

		final JList<String> list = new JList<String>(model);
		list.setSelectionMode(ListSelectionModel.SINGLE_SELECTION);
		list.setToolTipText("Choose Modpack Version");
		list.setBounds(10, 150, 414, 23);
		JScrollPane scrollPane = new JScrollPane(list);
		scrollPane.setBounds(10, 147, 414, 26);
		frmDownloadKingMammoth.getContentPane().add(scrollPane);

		JSeparator separator_3 = new JSeparator();
		separator_3.setBounds(10, 184, 414, 2);
		frmDownloadKingMammoth.getContentPane().add(separator_3);

		JButton btnNewButton = new JButton("Download");
		btnNewButton.addMouseListener(new MouseAdapter() {
			@Override
			public void mousePressed(MouseEvent arg0) {

				if (list.isSelectionEmpty() || textField.getText().isEmpty()) {

					new ChosenError();

				}

				else {

					currentSelection = list.getSelectedIndex();

					Release chosenFile = releases[currentSelection];

					frmDownloadKingMammoth.setVisible(false);
					EnchancedProgressBar.dir = textField.getText();
					EnchancedProgressBar.url = chosenFile.url;

					System.out.println("Loaded File");

					try {
						EnchancedProgressBar bar = new EnchancedProgressBar();
						bar.start();
						bar.backgroundTask();
					} catch (IOException e) {
						// TODO Auto-generated catch block
						e.printStackTrace();
					}

				}

			}
		});
		btnNewButton.setBackground(Color.LIGHT_GRAY);
		btnNewButton.setFocusable(false);
		btnNewButton.setBounds(10, 203, 131, 47);
		frmDownloadKingMammoth.getContentPane().add(btnNewButton);

		JButton btnCancel = new JButton("Cancel");
		btnCancel.setBackground(Color.LIGHT_GRAY);
		btnCancel.setFocusable(false);
		btnCancel.addMouseListener(new MouseAdapter() {
			@Override
			public void mousePressed(MouseEvent arg0) {
				frmDownloadKingMammoth.setVisible(false);
				frmDownloadKingMammoth.dispose();
				System.exit(0);
			}
		});
		btnCancel.setBounds(293, 203, 131, 47);
		frmDownloadKingMammoth.getContentPane().add(btnCancel);

		JButton btnSupportUs = new JButton("  Support Us");
		btnSupportUs.setBackground(Color.LIGHT_GRAY);
		btnSupportUs.setFocusable(false);
		btnSupportUs.setFont(new Font("Tahoma", Font.BOLD, 10));
		btnSupportUs.addMouseListener(new MouseAdapter() {
			@Override
			public void mousePressed(MouseEvent e) {

				String url = "https://www.patreon.com/kingmammoth";

				if (Desktop.isDesktopSupported()) {
					Desktop desktop = Desktop.getDesktop();
					try {
						desktop.browse(new URI(url));
					} catch (IOException | URISyntaxException e2) {
						// TODO Auto-generated catch block
						e2.printStackTrace();
					}
				} else {
					Runtime runtime = Runtime.getRuntime();
					try {
						runtime.exec("xdg-open " + url);
					} catch (IOException e1) {
						// TODO Auto-generated catch block
						e1.printStackTrace();
					}
				}
			}
		});
		btnSupportUs.setIcon(new ImageIcon(
				DownloadModpack.class.getResource("/king_mammoth_org/kingmammothlauncher/swing/patreon.png")));
		btnSupportUs.setBounds(151, 203, 131, 47);
		frmDownloadKingMammoth.getContentPane().add(btnSupportUs);

		JButton btnNewButton_1 = new JButton("Choose File");
		btnNewButton_1.setBackground(Color.LIGHT_GRAY);
		btnNewButton_1.setFocusable(false);
		btnNewButton_1.addMouseListener(new MouseAdapter() {
			@Override
			public void mousePressed(MouseEvent arg0) {
				try {
					UIManager.setLookAndFeel(UIManager.getSystemLookAndFeelClassName());
				} catch (ClassNotFoundException | InstantiationException | IllegalAccessException
						| UnsupportedLookAndFeelException e) {
					// TODO Auto-generated catch block
					e.printStackTrace();
				}
				JFileChooser chooser = new JFileChooser();
				chooser.setCurrentDirectory(new java.io.File("."));
				chooser.setDialogTitle("Choose Directory");
				chooser.setFileSelectionMode(JFileChooser.DIRECTORIES_ONLY);
				chooser.setAcceptAllFileFilterUsed(false);

				if (chooser.showOpenDialog(null) == JFileChooser.APPROVE_OPTION) {
					downloadDirectory = chooser.getSelectedFile();
					textField.setText(downloadDirectory.getAbsolutePath());

				} else {
					System.out.println("No Selection");
				}
			}
		});
		btnNewButton_1.setBounds(335, 109, 89, 23);
		frmDownloadKingMammoth.getContentPane().add(btnNewButton_1);
	}
}
